cmd/cli/job/root.go

```diff
@@ -2,23 +2,25 @@ package job
 
 import (
 	"github.com/bacalhau-project/bacalhau/cmd/util/hook"
+	"github.com/bacalhau-project/bacalhau/pkg/config"
+
 	"github.com/spf13/cobra"
 )
 
-func NewCmd() *cobra.Command {
+func NewCmd(cfg config.Context) *cobra.Command {
 	cmd := &cobra.Command{
 		Use:                "job",
 		Short:              "Commands to submit, query and update jobs.",
 		PersistentPreRunE:  hook.AfterParentPreRunHook(hook.RemoteCmdPreRunHooks),
 		PersistentPostRunE: hook.AfterParentPostRunHook(hook.RemoteCmdPostRunHooks),
 	}
 
-	cmd.AddCommand(NewDescribeCmd())
-	cmd.AddCommand(NewExecutionCmd())
-	cmd.AddCommand(NewHistoryCmd())
-	cmd.AddCommand(NewListCmd())
-	cmd.AddCommand(NewLogCmd())
-	cmd.AddCommand(NewRunCmd())
-	cmd.AddCommand(NewStopCmd())
+	cmd.AddCommand(NewDescribeCmd(cfg))
+	cmd.AddCommand(NewExecutionCmd(cfg))
+	cmd.AddCommand(NewHistoryCmd(cfg))
+	cmd.AddCommand(NewListCmd(cfg))
+	cmd.AddCommand(NewLogCmd(cfg))
+	cmd.AddCommand(NewRunCmd(cfg))
+	cmd.AddCommand(NewStopCmd(cfg))
```

Comment:
> recommendation is to create the client during the root's PersistentPreRunE and make it available to the command

This would require the repo to be created in the root command, which is something I feel we should avoid.

> Since not all commands need a client, there are two ways we can do that

I think these are worth exploring in a follow on, especially when it's no longer the case that different commands need different versions the client (v1 v v2) to operate.

Reply:
> This would require the repo to be created in the root command, which is something I feel we should avoid.

No we don't have to. Both options I explained don't require you to create a repo for all commands, but only for the commands with certain annotation of specific type

> I think these are worth exploring in a follow on

What we are doing now is not a good practice where we create a cfg, pass it to commands, and then modify it in preRun, where we can do most things in preRun. It is not right to push this now and then try to fix it later, and having two clients is not complicating things significantly